### PR TITLE
enh: target cmd - only register target flags with higher specificty

### LINKED ISSUE #195: target: only target flags which higher specificity shall be available for subcommands
**What would you like to be added**:

Currently the target flags `--garden`, `--project`, `--shoot`, `--seed`, `--control-plane` are available for the `target` and all of its subcommands.

Clarify if

1) we want to keep it this way and allow sth. like `target project myproject --garden mygarden` OR
2) the target flags should only be available if none of the target-subcommand is used (`target --garden mygarden`)
**3) only target flags with a higher specificity are allowed. So `target project myproject --shoot myshoot` would be ok but not `target project myproject --garden mygarden` would not.**

**Why is this needed**:

follow up for [#85](https://github.com/gardener/gardenctl-v2/issues/85)

### PR DESCRIPTION
**Pull request by sven-petersen**
**What this PR does / why we need it**:

see linked issue

**Which issue(s) this PR fixes**:
Fixes #195 

**Special notes for your reviewer**:

**Release note**:

```breaking user
Only allow target flags (`--garden`, ...) with lower specificity when using `target` command
```

### CODE REVIEW

**Comment by gardener-robot-ci-2** [comment]
Thank you @sven-petersen for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

**Review by petersutter — COMMENTED** [review]
nit

File: pkg/cmd/target/seed.go [hunk 1]
```diff
@@ -36,5 +38,11 @@ gardenctl target seed my-seed --garden my-garden`,
 
 	o.AddFlags(cmd.Flags())
 
+	manager, err := f.Manager()
+	utilruntime.Must(err)
+	manager.TargetFlags().AddGardenFlag(cmd.Flags())
+	manager.TargetFlags().AddProjectFlag(cmd.Flags())
```
Comment (petersutter) [inline pkg/cmd/target/seed.go:44]:
```suggestion
```
remove

File: pkg/cmd/target/view.go [hunk 2]
```diff
@@ -35,6 +37,11 @@ func NewCmdView(f util.Factory, ioStreams util.IOStreams) *cobra.Command {
 
 	o.AddFlags(cmd.Flags())
 
+	manager, err := f.Manager()
+	utilruntime.Must(err)
+	manager.TargetFlags().AddFlags(cmd.Flags())
```
Comment (petersutter) [inline pkg/cmd/target/view.go:42]: not sure if we need target flags on the view command

**Review by petersutter — APPROVED** [review]
/lgtm

File: pkg/cmd/target/control-plane.go [hunk 3]
```diff
@@ -36,5 +38,13 @@ gardenctl target control-plane --garden my-garden --project my-project --shoot m
 
 	o.AddFlags(cmd.Flags())
 
+	manager, err := f.Manager()
+	utilruntime.Must(err)
```
Comment (holgerkoser) [inline pkg/cmd/target/control-plane.go:42]: It is not acceptable to assert no erros here. We never have done this anywhere else.

File: internal/util/factory.go [hunk 4]
```diff
@@ -63,13 +66,19 @@ type FactoryImpl struct {
...
-	// TargetFlags can be used to completely override the target configuration
+	// targetFlags can be used to completely override the target configuration
 	// stored on the filesystem via a CLI flags.
-	TargetFlags target.TargetFlags
+	targetFlags target.TargetFlags
 }
 
 var _ Factory = &FactoryImpl{}
 
+func NewFactory() *FactoryImpl {
```
Comment (holgerkoser) [inline internal/util/factory.go:76]: Since this returns the strcut it should be named `NewFactoryImpl`?

File: pkg/flags/target_test.go [hunk 5]
```diff
@@ -263,10 +260,7 @@ var _ = Describe("Target flags", func() {
 		var factory *util.FactoryImpl
 
 		BeforeEach(func() {
-			factory = &util.FactoryImpl{
-				TargetFlags: targetFlags,
-				ConfigFile:  configFile,
-			}
+			factory = util.NewFactory()
```
Comment (holgerkoser) [inline pkg/flags/target_test.go:263]:
```
factory = util.NewFactory()
factory.ConfigFile = configFile
```

File: internal/util/factory.go [hunk 6]
```diff
@@ -63,13 +66,19 @@ type FactoryImpl struct {
...
+	return &FactoryImpl{
+		targetFlags: target.NewTargetFlags("", "", "", "", false),
+	}
+}
```
Comment (holgerkoser) [inline internal/util/factory.go:80]:
```go
func NewFactoryImpl(tf target.TargetFlags) *FactoryImpl {
        var targetFlags target.TargetFlags
        if (tf != nil) {
           targetFlags = tf
        } else {
           targteFlags = target.NewTargetFlags("", "", "", "", false)
        }
	return &FactoryImpl{
		targetFlags,
	}
}
```

File: pkg/flags/target_test.go (hunk 5 above)
Comment (holgerkoser) [reply pkg/flags/target_test.go]: We agreed, that we don't need it and for new tests we can use the mockFactory.

**Review by holgerkoser — APPROVED** [review]
/lgtm

**Pull request merged**